File: app/src/main/java/com/esorokin/lantector/ui/plugins/ToolbarPlugin.java
package com.esorokin.lantector.ui.plugins;

import android.support.annotation.NonNull;
import android.support.annotation.StringRes;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.text.TextUtils;
import android.view.View;
import android.view.ViewGroup;

import com.esorokin.lantector.ui.plugins.base.BaseDependencyPlugin;

public class ToolbarPlugin extends BaseDependencyPlugin<AppCompatActivity> {
	@NonNull
	private String title;

	public ToolbarPlugin(AppCompatActivity delegate, @StringRes int titleRes) {
		this(delegate, delegate.getString(titleRes));
	}

	public ToolbarPlugin(AppCompatActivity delegate, @NonNull String title) {
		super(delegate);
		this.title = title;
	}

	public ToolbarPlugin(AppCompatActivity delegate) {
		super(delegate);
		this.title = "";
	}

	@Override
	public void onViewCreated(final View view) {
		super.onViewCreated(view);
		Toolbar toolbar = getToolbar(view);
		if (toolbar != null) {
			getDependency().setSupportActionBar(toolbar);
			if (!TextUtils.isEmpty(title)) {
				getDependency().setTitle(title);
			}
		}
	}

	public void setTitle(@NonNull String title) {
		this.title = title;
		getDependency().setTitle(title);
	}

	private Toolbar getToolbar(View view) {
		if (view instanceof Toolbar) {
			return (Toolbar) view;
		}
		if (view instanceof ViewGroup) {
			for (int i = 0; i < ((ViewGroup) view).getChildCount(); i++) {
				Toolbar toolbar = getToolbar(((ViewGroup) view).getChildAt(i));
				if (toolbar != null) {
					return toolbar;
				}
			}
		}
		return null;
	}
}

File: app/src/main/java/com/esorokin/lantector/ui/plugin/ToolbarPlugin.kt
package com.esorokin.lantector.ui.plugin

import android.support.annotation.StringRes
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.Toolbar
import android.view.View
import android.view.ViewGroup
import com.esorokin.lantector.ui.plugin.base.BasePlugin

open class ToolbarPlugin(protected val activityProvider: () -> AppCompatActivity,
                         @StringRes val titleRes: Int = 0,
                         private var title: String? = null) : BasePlugin() {
    private val toolbar: Toolbar? by lazy { getToolbar(activityProvider.invoke().window.decorView) }

    override fun onViewCreated(view: View) {
        super.onViewCreated(view)

        val activity = activityProvider.invoke()

        if (title.isNullOrEmpty().and(titleRes != 0)) {
            title = activity.getString(titleRes)
        }

        toolbar?.let {
            if (activity.supportActionBar == null) {
                activity.setSupportActionBar(it)
            }

            if (!title.isNullOrEmpty()) {
                it.title = title
                activity.title = title
            }
        }
    }

    fun changeTitle(title: String) {
        this.title = title
        toolbar?.title = title
        activityProvider.invoke().title = title
    }

    private fun getToolbar(view: View): Toolbar? {
        if (view is Toolbar) {
            return view
        }

        if (view is ViewGroup) {
            (0..view.childCount - 1)
                    .mapNotNull { getToolbar(view.getChildAt(it)) }
                    .forEach { return it }
        }

        return null
    }
}
